### PR TITLE
Add nvim floating window support (replaces #3314)

### PR DESCRIPTION
**Pull request by kevinclark**
This continues the work begun in #3341 by @jargonzombies and responds to the feedback from @w0rp. In particular, it splits out configuration of the individual floating-window usecases and pulls that code into a separate file.

Known issues:
  * ~~Needs tests. Since it's my first time working in viml, I wanted to make sure the basics were right before digging in here.~~ Done and passing as of 60287b0.
  * ~~When floating preview coupled with g:ale_cursor_detail, the window is redrawn in a tight loop. Presumably in the non-floating version, moving into the preview window (via `stay_here`) keeps this from happening. But `stay_here` doesn't make sense in a floating-window world, so we may need another fix.~~ This now fixed in c39a27c.

As far as I'm concerned this is ready to go. Just waiting on review and a merge.

### CODE REVIEW

**Comment by kevinclark** [comment]
@w0rp I'm going to work on tests next. I'd appreciate some direction re: the noted `stay_here` issue. @jargonzombies initial patch moved control into the floating preview window if `stay_here` was set, but it seems pretty jarring that control is suddenly in this tiny float. May or may not be obvious to a user that the float is a real buffer that can be window-closed or moved out of. Is there another way to stop the loop until `CursorMoved,TabLeave,WinLeave,InsertEnter`?

**Comment by kevinclark** [comment]
Currently considering progress on this blocked by #3448 since I can't run passing tests locally. Working on that first.

**Comment by kevinclark** [comment]
Just an update here - I've got a PR in to fix #3448. It's waiting on review/feedback/merge. I'm happy to write the tests to finish this up, but wanted a basic "this seems like the right thing" sanity check before I go learn vader.

**Comment by kevinclark** [comment]
Still waiting on any kind of feedback. 🤷‍♂️

Just maintaining my own branch that I’m using daily in the meantime. Works fine afaict.

**Comment by cj** [comment]
@kevinclark with Rubocop, the floating preview is missing the option that caused the error:  
CleanShot 2020-12-29 at 14 06 51

**Comment by kevinclark** [comment]
@cj I’ll take a look. Would you mind supplying any ale options you’ve got set so I can repro?

**Comment by cj** [comment]
@kevinclark of course - https://gist.github.com/cj/39eeb5de63691f114f50832b444fdc3f, let me know if you need any thing else!

**Comment by kevinclark** [comment]
@cj Thanks! And just in case it matters (I don’t think it should) - what version of NeoVim?

**Comment by cj** [comment]
```
NVIM v0.5.0-dev+950-g5ce328df4
Build type: Release
LuaJIT 2.1.0-beta3
```

**Comment by kevinclark** [comment]
@cj While this shows up because of the floating code, it's actually unrelated. Looks like the preview window paths don't use the same formatting as the cursor detail paths.

In any case, it's a small change to make it consistent. I've tossed up a preliminary PR [here](https://github.com/dense-analysis/ale/pull/3505). Will work up tests shortly.

**Comment by davidatbu** [comment]
Thanks for this @kevinclark ! 

So another nice-to-have would be `g:ale_floating_cursor` (in the vein of `g:ale_virtualtext_cursor`), which would automatically start a floating win when the cursor is on a line with an error/warning. 

But I'm already very grateful for this PR! Maybe someone else (like me) should take this on after we get this PR merged in it's current form.

**Comment by kevinclark** [comment]
@davidatbu I think what you're looking for already exists. If you turn on `g:ale_cursor_detail` and `g:ale_floating_preview` then the cursor detail will use the floating window instead of the split preview. I'm running with this setting.

There is a slight difference in output, which @cj points out and I refer to in #3505  where the `detail` vs `text` output of the locline entry is used, but it was useful enough for me that I hadn't noticed.

**Comment by kevinclark** [comment]
Open question (possibly for a follow up PR): when using ALEDocumentation, should floating windows be an option?

**Comment by kevinclark** [comment]
Ok, tests are done. Just waiting on a review and merge. /cc @w0rp

**Comment by kevinclark** [comment]
@hsanson Looks like you're the person actually merging PRs. Can you take a look at this?

**Comment by hsanson** [comment]
@kevinclark I would love to be able review this but my knowledge of ALE core parts is not enough to review this properly. 

From what I understand this adds three configuration variables that control if the ALEInfo and Hover information is displayed in a preview window in a split or as a floating window (only if neovim is used).

I set these on my configuration:

```
let g:ale_hover_cursor = 1
let g:ale_hover_to_preview = 1
let g:ale_float_preview = 1
let g:ale_floating_preview = 1
```

And in a simple ruby file with solargraph/rubocop linters all I get is an error message:

```
 [rubocop] (Style/FrozenStringLiteralComment) - Missing frozen string literal comment.
Error detected while processing function <SNR>114_NeoVimCallback[29]..<lambda>17[1]..<SNR>110_ExitCallback[28]..<SNR>111_HandleExit[30]..al
e#engine#HandleLoclist[40]..ale#engine#SetResults[26]..ale#cursor#EchoCursorWarning[34]..<SNR>143_ShowCursorDetailForItem[8]..ale#floating_
preview#Show:
line   45:
E5555: API call: Vim:E315: ml_get: invalid lnum: 2
```

My full ALEInfo:

```
 Current Filetype: ruby
Available Linters: ['brakeman', 'debride', 'rails_best_practices', 'reek', 'rubocop', 'ruby', 'solargraph', 'sorbet', 'standardrb']
   Linter Aliases:
'sorbet' -> ['srb']
  Enabled Linters: ['rubocop', 'ruby', 'solargraph']
  Ignored Linters: []
 Suggested Fixers: 
  'remove_trailing_lines' - Remove all blank lines at the end of a file.
  'rubocop' - Fix ruby files with rubocop --auto-correct.
  'rufo' - Fix ruby files with rufo
  'sorbet' - Fix ruby files with srb tc --autocorrect.
  'standardrb' - Fix ruby files with standardrb --fix
  'trim_whitespace' - Remove all trailing whitespace characters at the end of every line.
 Linter Variables:

let g:ale_ruby_rails_best_practices_executable = 'rails_best_practices'
let g:ale_ruby_rails_best_practices_options = ''
let g:ale_ruby_rubocop_executable = 'rubocop'
let g:ale_ruby_rubocop_options = ''
let g:ale_ruby_ruby_executable = 'ruby'
let g:ale_ruby_solargraph_executable = 'solargraph'
let g:ale_ruby_solargraph_options = {}
 Global Variables:

let g:ale_cache_executable_check_failures = v:null
let g:ale_change_sign_column_color = 0
let g:ale_command_wrapper = ''
let g:ale_completion_delay = 100
let g:ale_completion_enabled = 0
let g:ale_completion_max_suggestions = 50
let g:ale_disable_lsp = 0
let g:ale_echo_cursor = 1
let g:ale_echo_msg_error_str = ''
let g:ale_echo_msg_format = '%severity% [%linter%] (%code%) - %s'
let g:ale_echo_msg_info_str = ''
let g:ale_echo_msg_warning_str = ''
let g:ale_enabled = 1
let g:ale_fix_on_save = 0
let g:ale_fixers = {'bib': ['bibclean'], 'openapi': ['prettier'], 'kotlin': ['ktlint'], '*': ['remove_trailing_lines', 'trim_whitespace'], 'yaml': ['prettier'], 'ruby': ['rubocop'], 'python': ['yapf']}
let g:ale_history_enabled = 1
let g:ale_history_log_output = 1
let g:ale_keep_list_window_open = 0
let g:ale_lint_delay = 200
let g:ale_lint_on_enter = 1
let g:ale_lint_on_filetype_changed = 1
let g:ale_lint_on_insert_leave = 1
let g:ale_lint_on_save = 1
let g:ale_lint_on_text_changed = 'normal'
let g:ale_linter_aliases = {'asciidoctor': 'asciidoc'}
let g:ale_linters = {'mail': ['proselint', 'write-good'], 'asciidoc': ['vale'], 'text': ['proselint', 'write-good'], 'latex': ['proselint', 'chktex', 'lacheck'], 'kotlin': ['ktlint', 'languageserver'], 'go': ['gofmt', 'golint', 'go vet', 'golangserver'], 'markdown': ['markdownlint', 'vale'], 'yaml': ['yamllint'], 'python': ['flake8', 'pylint', 'pyright'], 'tex': ['proselint', 'chktex', 'lacheck'], 'vim': ['vint'], 'bib': ['bibclean'], 'xml': ['android'], 'java': ['android', 'checkstyle', 'javalsp'], 'javascript': ['eslint'], 'plaintex': ['proselint', 'chktex', 'lacheck'], 'help': [], 'openapi': ['yamllint', 'ibm-validator'], 'groovy': ['android'], 'dockerfile': ['dockerfile_lint'], 'ruby': ['solargraph', 'rubocop', 'ruby']}
let g:ale_linters_explicit = 0
let g:ale_linters_ignore = {}
let g:ale_list_vertical = 0
let g:ale_list_window_size = 10
let g:ale_loclist_msg_format = '%severity% [%linter%] (%code%) - %s'
let g:ale_lsp_root = {}
let g:ale_max_buffer_history_size = 20
let g:ale_max_signs = -1
let g:ale_maximum_file_size = v:null
let g:ale_open_list = 0
let g:ale_pattern_options = {'.gitlab-ci\.yml$': {'ale_linters': ['gitlablint', 'yamllint']}}
let g:ale_pattern_options_enabled = v:null
let g:ale_set_balloons = 0
let g:ale_set_highlights = 1
let g:ale_set_loclist = 1
let g:ale_set_quickfix = 0
let g:ale_set_signs = 1
let g:ale_sign_column_always = 0
let g:ale_sign_error = ''
let g:ale_sign_info = ''
let g:ale_sign_offset = 1000000
let g:ale_sign_style_error = ''
let g:ale_sign_style_warning = ''
let g:ale_sign_warning = ''
let g:ale_sign_highlight_linenrs = 0
let g:ale_statusline_format = v:null
let g:ale_type_map = {}
let g:ale_use_global_executables = v:null
let g:ale_virtualtext_cursor = 1
let g:ale_warn_about_trailing_blank_lines = 1
let g:ale_warn_about_trailing_whitespace = 1
  Command History:

(executable check - success) solargraph
(executable check - success) rubocop
(finished - exit code 1) ['/bin/bash', '-c', '''rubocop'' --format json --force-exclusion  --stdin ''/home/ryujin/Projects/HelloWorld/ruby/array.rb'' < ''/tmp/nvimtxtmel/2/array.rb''']

<<<OUTPUT STARTS>>>
{"metadata":{"rubocop_version":"1.7.0","ruby_engine":"ruby","ruby_version":"2.7.0","ruby_patchlevel":"0","ruby_platform":"x86_64-linux-gnu"},"files":[{"path":"array.rb","offenses":[{"severity":"convention","message":"Missing frozen string literal comment.","cop_name":"Style/FrozenStringLiteralComment","corrected":false,"correctable":true,"location":{"start_line":1,"start_column":1,"last_line":1,"last_column":1,"length":1,"line":1,"column":1}}]}],"summary":{"offense_count":1,"target_file_count":1,"inspected_file_count":1}}
<<<OUTPUT ENDS>>>

(executable check - success) ruby
(finished - exit code 0) ['/bin/bash', '-c', '''ruby'' -w -c -T1 ''/tmp/nvimtxtmel/3/array.rb''']

<<<OUTPUT STARTS>>>
ruby: warning: ruby -T will be removed in Ruby 3.0
<<<OUTPUT ENDS>>>

(started) ['/bin/bash', '-c', '''solargraph'' stdio']
```

**Comment by kevinclark** [comment]
@hsanson Odd, I can't seem to reproduce. FYI, `g:ale_float_preview` doesn't exist at this point.

Could you provide the file that's causing that error?

I can't seem to get solargraph to install (compile errors) but rubocop errors don't cause this in my tree with your options.

**Comment by hsanson** [comment]
Sample files:

array.rb:

```ruby
a = { a: 1, b: 2 }
puts a
```

Gemfile:

```Gemfile
source "https://rubygems.org"

gem "solargraph"
gem "standardrb"

gem "http"
```

Both located on the same directory and nevim CD'd into it. Using latest neovim and ale from git master.

**Comment by kevinclark** [comment]
@hsanson Huh. That works fine for me. To clarify - when you say ale from git master - do you mean the ale from my branch? I hadn't merged from ale master in a little while, but I have now and it's still working fine. A little repetitive with `g:ale_hover_cursor` on, but working correctly AFAICT.

**Comment by hsanson** [comment]
Sorry, not master... I am using your branch:

```sh
git checkout -b kevinclark-nvim-floating master
git pull git://github.com/kevinclark/ale.git nvim-floating
```

```sh
nvim -v
NVIM v0.5.0-888-g8b2887bd5
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/cc -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/home/hs/Apps/neovim.git/build/config -I/home/hs/Apps/neovim.git/src -I/home/hs/Apps/neovim.git/.deps/usr/include -I/usr/include -I/home/hs/Apps/neovim.git/build/src/nvim/auto -I/home/hs/Apps/neovim.git/build/include
Compiled by hs@ex

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```

**Comment by kevinclark** [comment]
@hsanson Got it. I'm on nvim 0.4.4 - I'll move to 0.5 and see if there's a change. Appreciate the responsiveness!

**Comment by kevinclark** [comment]
@hsanson Works with nightly 0.5 too. How odd. Are you seeing any test failures when you run them?

**Comment by hsanson** [comment]
Ok, so I tried with a minimal vim configuration and worked fine. After commenting line by line my config I found this particular setting causes the issue to occur:

```
set scrolloff=9999      " Always keep the cursor at the center of window.
```

Commenting this line solves the issue and the PR seems to work as expected. Unfortunately I am inclined to believe this configuration is common in the wild (see https://vim.fandom.com/wiki/Keep_your_cursor_centered_vertically_on_the_screen) so this PR can potentially break things to many people as it did for me.

Maybe is a bug in neovim itself that ALE has no way of working around. Would be good if someone can test this in vim to see if the same problem occurs.

**Comment by kevinclark** [comment]
@hsanson Thanks so much for tracking that down - I'll do some digging and see if I can't isolate the problem.

**Comment by kevinclark** [comment]
@hsanson So, with nightly nvim I can't get the branch to fail setting scroll off and I'm seeing test pass with this diff:

```diff
diff --git a/test/test_floating_preview.vader b/test/test_floating_preview.vader
index 43415556..46213641 100644
--- a/test/test_floating_preview.vader
+++ b/test/test_floating_preview.vader
@@ -1,4 +1,7 @@
 Before:
+  Save scrolloff
+  set scrolloff=9999
+
   let g:ale_floating_preview = 0
   let g:ale_hover_to_floating_preview = 0
   let g:ale_detail_to_floating_preview = 0
```

Kind of stumped. Does it blow up on that two line ruby file with that scrolloff or does it need to be something larger?

**Comment by hsanson** [comment]
@kevinclark I updated my neovim installation to latest master and the issue I had no longer occurs. There are many changes between my previous version and the latest one so is difficult to tell what was the problem.

I think this works as expected and can be merged but I also think @w0rp should check on this and merge it.

**Review by hsanson — COMMENTED** [review]
Changes look good and tested it works as expected using latest neovim build.

**Comment by kevinclark** [comment]
@hsanson Thank you for the help!

File: autoload/ale/floating_preview.vim
Comment (w0rp) [inline autoload/ale/floating_preview.vim:15]: We should prefix groups we create with `ALE` so we don't conflict with other groups.

File: doc/ale.txt
Comment (w0rp) [inline doc/ale.txt:1195]: The default is `0`.

**Review by w0rp — CHANGES_REQUESTED** [review]
I just have a couple of comments here. Otherwise this is fine.

Sorry for not looking anything to do with ALE for a while. Given the times we are in and my day job, I haven't felt like doing too much with ALE for now.

**Comment by w0rp** [comment]
@hsanson GitHub really needs a private message system some day. I've given you Admin permissions for the ALE repo, so if you couldn't merge something before, you definitely should be able to merge things now.

**Comment by kevinclark** [comment]
@w0rp @hsanson I've made those changes. Let me know if there's anything else needed to get this in.

Also - it'd make work going forward a lot easier if tests were portable. In master right now some of the testing tooling is from the host system instead of the docker container. I've fixed that in #3471. If I could get some attention there, I'd appreciate it.

**Review by w0rp — APPROVED** [review]

**Comment by kevinclark** [comment]
Awesome, thank you!

**Comment by axeldetlofsson** [comment]
Great work! I just have a question, is the window ment to float at the sime location as the old preview window? Or pop up next to cursor?

image

**Comment by puresick** [comment]
Thank you for this feature PR - Great work!

**Pull request merged**
